Knative Kafka Channel Integration (#6985)
kyma-project/kyma: docs/event-bus/03-04-knative-eventing-mesh.md

For example, if you have used `gateway.example.cx/v1/events` so far, use `gateway.example.cx/events` to make sure you work with the new Eventing Mesh. 

>**NOTE:** The HTTP source adapter only accepts Events compliant with the [CloudEvents 1.0 specification](https://github.com/cloudevents/spec/blob/v1.0/spec.md).

## Other Channel options

> By default Kyma comes with Natss but Knative eventing allows to exchange the default channel implementation. It even allows to have multiple channel implementations simultaneously.

Author: Thanks for the link :)

> By default Kyma comes with Natss but Knative eventing allows to exchange the default channel implementation. It even allows to have multiple channel implementations simultaneously.

Author: `implement multiple channels simultaneously.` => I think it should be `use`, because in this documentation we only care about using channels, not implementing them. I would take your suggestion but with the word change. What do you think?

> The default channel implementation can be changed during installation using an installation-override like this:

Reviewer: ```suggestion
You can [override](root/kyma/#configuration-helm-overrides-for-kyma-installation) the default channel implementation during installation like this:
```
I would make it more direct - shorten the sentence, use the active voice (we usually go for active voice whenever possible), and just write `override` directly. I also added a link, in case somebody doesn't know how to override stuff.

Author: A leading slash is missing for `root`. I will pick your suggestion and change it afterwards

```bash
$ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: knative-eventing-overrides
  namespace: kyma-installer
  labels:
    installer: overrides
    component: knative-eventing
    kyma-project.io/installation: ""
data:
  knative-eventing.channel.default.apiVersion: knativekafka.kyma-project.io/v1alpha1
  knative-eventing.channel.default.kind: KafkaChannel
EOF
```

> In this example the default channel is set to Kafka.

Reviewer: ```suggestion
In this example, the default channel is set to Kafka.
```
Just a comma :)

### Kafka

> There is a Knative compatible [Kafka channel implementation](https://github.com/kyma-incubator/knative-kafka) which can be used for more production-ready eventing workloads but is still in alpha state.

Reviewer: ```suggestion
You can use a Knative-compatible [Kafka channel implementation](https://github.com/kyma-incubator/knative-kafka) for more production-ready eventing workloads. 
```
I removed the mention of alpha state because you already added a note below :) I also changed the voice to active.

> >**NOTE:** Kafka Channel integration is in alpha version. Use it only for testing purposes.

Reviewer 2: how can it be production ready and for test purposes only?

Author: Please check again

> The Knative channel implementation supports [these providers](https://github.com/kyma-incubator/knative-kafka/blob/9eb3fa3f6e67ffc80b162d2ef4c8a8a3942d9c5f/resources/README.md#kafka-providers):

Reviewer: Should we write `Knative-Kafka channel`? or Knative is enough? 
```suggestion
The Knative channel implementation supports [the following providers](https://github.com/kyma-incubator/knative-kafka/blob/9eb3fa3f6e67ffc80b162d2ef4c8a8a3942d9c5f/resources/README.md#kafka-providers). Use the links to set up a Kafka cluster.
```

Author: I think Knative alone is not enough, a Knative implementation could be a source, a channel, a trigger, a subscription, and lot of other objects. I would say `Knative-Kafka channel`. What do you think?

> 1. [Azure Event Hubs](https://azure.microsoft.com/en-us/services/event-hubs/)

Reviewer: ```suggestion
* [Azure Event Hubs](https://azure.microsoft.com/en-us/services/event-hubs/)
```
For this one and the following two points, I changed the numbering to bullet points - it's not a procedure, so it's better to use regular bullet points.

> 2. [Confluent Cloud](https://www.confluent.io/confluent-cloud)

Reviewer: ```suggestion
* [Confluent Cloud](https://www.confluent.io/confluent-cloud)
```

> 3. [Standard Kafka installation with no special authorization required](https://kafka.apache.org/quickstart)

Reviewer: ```suggestion
* [Standard Kafka installation with no special authorization required](https://kafka.apache.org/quickstart)
```

> Please follow any of the links above on how to setup a Kafka cluster.

Reviewer: ```suggestion

```
I moved this sentence up.

> Before starting the Kyma installation, the connection between Kyma and the Kafka cluster needs to be configured. This can be done via an installation override like this:

Reviewer: ```suggestion
Before starting the Kyma installation, configure the connection between Kyma and the Kafka cluster. To do so, use the installation override:
```

```bash
$ export $kafkaBrokers=<todo user>
$ export $kafkaNamespace=<todo user>
$ export $kafkaPassword=<todo user>
$ export $kafkaUsername=<todo user>
$ export $kafkaProvider=<local|azure|confluent>

$ cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: knative-kafka-overrides
  namespace: kyma-installer
  labels:
    installer: overrides
    component: knative-eventing-channel-kafka
    kyma-project.io/installation: ""
type: Opaque    
stringData:
  kafka.brokers: $kafkaBrokers
  kafka.namespace: $kafkaNamespace
  kafka.password: $kafkaPassword
  kafka.username: $kafkaUsername
  kafka.secretName: knative-kafka
  environment.kafkaProvider: $kafkaProvider
EOF
```

> >**NOTE:** For other options, check this [link](https://github.com/kyma-incubator/knative-kafka/blob/master/resources/knative-kafka/values.yaml).

Reviewer: ```suggestion
>**NOTE:** For additional values, see [this](https://github.com/kyma-incubator/knative-kafka/blob/master/resources/knative-kafka/values.yaml) file.
```

> Now that the installation has been customized, the Kyma installation can be triggered. 

Reviewer: ```suggestion
Once the installation has been customized, you can trigger the Kyma installation.
```
Active voice :)  Also, I assume that Kyma installation is triggered in the "standard" way if the override is configured before it was ever installed? Should we also mention that if the overrides are applied in the runtime, the user should run `kubectl label installation/kyma-installation action=install` ?

> You can install Kyma with a custom component (`knative-eventing-channel-kafka` for the component and `knative-eventing-channel-kafka-test` for the TestDefinition) by following these [instructions](/root/kyma/#configuration-custom-component-installation).

Reviewer: ```suggestion
You can install Kyma with a `knative-eventing-channel-kafka` custom component by following these [instructions](/root/kyma/#configuration-custom-component-installation). The TestDefinition for the component is `knative-eventing-channel-kafka-test`.
```
Question: if we install the component, the override is not necessary? How can the component be used?

Author: The component can't be used without an installer override, because the kafka channel controllers needs to know how to connect to the Azure Eventhubs Namespace. The secret for the controller could also be created after the installation. This would require a restart of the controller however. We could also document this approach if you like. Let me know.
